Notebooks protobuildings (#447)
BuildingSync/schema: docs/notebooks/bsync_examples/Primary-School-Levels1-2.md

Follow the instructions in the [README.md](../README.md) file to run this in Jupyter Lab/Notebook.

## 1. Pre-requisites

This notebook assumes that you have installed bsyncpy and all of its dependencies.

## 2. Useful information

This notebook uses resources from:

- the [DOE Commercial Prorotype buildings](https://www.energycodes.gov/prototype-building-models) repository
- ASHRAE [Standard 211](https://www.ashrae.org/technical-resources/bookstore/standards-180-and-211) for formal guidelines on energy audits
- the [BuildingSync schema documentation](https://buildingsync.net/dictionary/2.4.0/)
- the [notebook included in the bsync examples folder](./Small-Office-Level-1.md) for a comprehensive explanation on basic BuildingSync and bsyncpy concepts

## 3. Overview

This notebook is broken down into the following sections:

1) Setting up the root, or the base, of your BuildingSync file

2) Adding the information required for a Level 1 energy audit

3) Adding the information required for a Level 2 energy audit

4) Adding the information required for a Level 3 energy audit (WIP)

> 5) Generating a BuildingSync file and verifying it against version 2.4.0 of the BuildingSync schema

Reviewer: ```suggestion
4) Generating a BuildingSync file and verifying it against version 2.4.0 of the BuildingSync schema
```

```python
# we add the facilities (facs) to the root
root += facs
# and then we add Facility-1 (f1) to the facilities (facs)
facs += f1
# and we continue adding children to the parent elements, following the tree's branches all the way down to our building
f1 += sites
sites += s1
s1 += bldgs
bldgs += b1
```

And we verify the output of what we just did:

```python
pretty_print(root)
```

    <BuildingSync>
      <Facilities>
        <Facility ID="Facility-1">
          <Sites>
            <Site ID="Site-1">
              <Buildings>
                <Building ID="Building-Primary-School-Prototype"/>
              </Buildings>
            </Site>
          </Sites>
        </Facility>
      </Facilities>
    </BuildingSync>
    

This looks exactly like it was supposed to look. If you needed to add a different site, you would just need to add a child to the appropriate parent, in that case a Site to a group of Sites.
Now, let us carry on and start the Level 1 audit.

## 5. Level 1 audit

We refer to ASHRAE Standard 211, section 5.3, relevant to the Level 1 energy audit procedures.
For this audit, we will need to report:

> 1) a facility description (section 6.1.1.1)

Reviewer: ```suggestion
1) a facility description (section 6.1.1.1) <br>
```

> 2) the historical energy use (section 6.1.1.2)

Reviewer: ```suggestion
2) the historical energy use (section 6.1.1.2) <br>
```

> 3) a benchmark of the site EUI (section 6.1.1.3)

Reviewer: ```suggestion
3) a benchmark of the site EUI (section 6.1.1.3) <br>
```

> 4) the target and estimate savings (section 6.1.1.4)

Reviewer: ```suggestion
4) the target and estimate savings (section 6.1.1.4) <br>
```

> 5) low-cost and no-cost energy efficiency measure recommendations (section 6.1.1.5)

Reviewer: ```suggestion
5) low-cost and no-cost energy efficiency measure recommendations (section 6.1.1.5) <br>
```

> 6) potential capital energy efficiency measure recommendations (section 6.1.1.6)

Reviewer: ```suggestion
6) potential capital energy efficiency measure recommendations (section 6.1.1.6) <br>
```

```

#### 5.2.2 Mechanical Room

```python
# create a new section
mech_sec = bsync.Sections.Section(ID="Mechanical-room")
mech_sec += bsync.SectionType('Space function')
mech_sec += bsync.OccupancyClassification("Mechanical room")
mech_sec += bsync.OriginalOccupancyClassification("Education-Primary")
sections += mech_sec

# add the floor area

# 5.3.4.a Gross floor area for the section (we also add conditioned for 5.3.4.f)
mech_fas = bsync.FloorAreas(
    bsync.FloorArea(
        bsync.FloorAreaType("Gross"),
        bsync.FloorAreaValue(2712.51)
    ),
    bsync.FloorArea(
        bsync.FloorAreaType("Conditioned"),
        bsync.FloorAreaValue(2712.51)
    )
)

# 5.3.4.b and 5.3.4.c are asking for typical operating hours for the section
mech_tous = bsync.TypicalOccupantUsages()
mech_tou_hpw = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Hours per week"),
    bsync.TypicalOccupantUsageValue(30.0)
)
mech_tou_wpy = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Weeks per year"),
    bsync.TypicalOccupantUsageValue(50.)
)
mech_tous += mech_tou_hpw
mech_tous += mech_tou_wpy

# 5.3.4.d looks for normal / design or peak occupancy.

> # Here, we take values from the IDF as the building has been modelled: with no occupants in the mech room, bathroom or corridors.

Reviewer: ```suggestion
# Here, we take values from the IDF as the building has been modeled: with no occupants in the mech room, lobby, bathroom or corridors.
```

mech_sec += mech_occ_levels
```

#### 5.2.3 Library

```python
# create a new section
lib_sec = bsync.Sections.Section(ID="Library")
lib_sec += bsync.SectionType('Space function')
lib_sec += bsync.OccupancyClassification("Media center")
lib_sec += bsync.OriginalOccupancyClassification("Education-Primary")
sections += lib_sec

# add the floor area

# 5.3.4.a Gross floor area for the section (we also add conditioned for 5.3.4.f)
lib_fas = bsync.FloorAreas(
    bsync.FloorArea(
        bsync.FloorAreaType("Gross"),
        bsync.FloorAreaValue(4294.80)
    ),
    bsync.FloorArea(
        bsync.FloorAreaType("Conditioned"),
        bsync.FloorAreaValue(4294.80)
    )
)

# 5.3.4.b and 5.3.4.c are asking for typical operating hours for the section
lib_tous = bsync.TypicalOccupantUsages()
lib_tou_hpw = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Hours per week"),
    bsync.TypicalOccupantUsageValue(45.0)
)
lib_tou_wpy = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Weeks per year"),
    bsync.TypicalOccupantUsageValue(50.)
)
lib_tous += lib_tou_hpw
lib_tous += lib_tou_wpy

# 5.3.4.d looks for normal / design or peak occupancy.

> # Here, we take values from the IDF as the building has been modelled: with no occupants in the mech room, bathroom or corridors.

Reviewer: ```suggestion
# Here, we take values from the IDF as the building has been modeled: with no occupants in the mech room, lobby bathroom or corridors.
```

```

#### 5.2.4 Kitchen

```python
# create a new section
kitchen_sec = bsync.Sections.Section(ID="Kitchen")
kitchen_sec += bsync.SectionType('Space function')
kitchen_sec += bsync.OccupancyClassification("Kitchen")
kitchen_sec += bsync.OriginalOccupancyClassification("Education-Primary")
sections += kitchen_sec

# add the floor area

# 5.3.4.a Gross floor area for the section (we also add conditioned for 5.3.4.f)
kitchen_fas = bsync.FloorAreas(
    bsync.FloorArea(
        bsync.FloorAreaType("Gross"),
        bsync.FloorAreaValue(1808.33)
    ),
    bsync.FloorArea(
        bsync.FloorAreaType("Conditioned"),
        bsync.FloorAreaValue(1808.33)
    )
)

# 5.3.4.b and 5.3.4.c are asking for typical operating hours for the section
kitchen_tous = bsync.TypicalOccupantUsages()
kitchen_tou_hpw = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Hours per week"),
    bsync.TypicalOccupantUsageValue(45.0)
)
kitchen_tou_wpy = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Weeks per year"),
    bsync.TypicalOccupantUsageValue(50.)
)
kitchen_tous += kitchen_tou_hpw
kitchen_tous += kitchen_tou_wpy

# 5.3.4.d looks for normal / design or peak occupancy.

> # Here, we take values from the IDF as the building has been modelled: with no occupants in the mech room, bathroom or corridors.

Reviewer: ```suggestion
# Here, we take values from the IDF as the building has been modeled: with no occupants in the mech room, lobby, bathroom or corridors.
```

kitchen_sec += kitchen_occ_levels
```

#### 5.2.5 Gym

```python
# create a new section
gym_sec = bsync.Sections.Section(ID="Gym")
gym_sec += bsync.SectionType('Space function')
gym_sec += bsync.OccupancyClassification("Sport play area")
gym_sec += bsync.OriginalOccupancyClassification("Education-Primary")
sections += gym_sec

# add the floor area

# 5.3.4.a Gross floor area for the section (we also add conditioned for 5.3.4.f)
gym_fas = bsync.FloorAreas(
    bsync.FloorArea(
        bsync.FloorAreaType("Gross"),
        bsync.FloorAreaValue(3842.71)
    ),
    bsync.FloorArea(
        bsync.FloorAreaType("Conditioned"),
        bsync.FloorAreaValue(3842.71)
    )
)

# 5.3.4.b and 5.3.4.c are asking for typical operating hours for the section
gym_tous = bsync.TypicalOccupantUsages()
gym_tou_hpw = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Hours per week"),
    bsync.TypicalOccupantUsageValue(25.0)
)
gym_tou_wpy = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Weeks per year"),
    bsync.TypicalOccupantUsageValue(50.)
)
gym_tous += gym_tou_hpw
gym_tous += gym_tou_wpy

# 5.3.4.d looks for normal / design or peak occupancy.

> # Here, we take values from the IDF as the building has been modelled: with no occupants in the mech room, bathroom or corridors.

Reviewer: ```suggestion
# Here, we take values from the IDF as the building has been modeled: with no occupants in the mech room, lobby, bathroom or corridors.
```

classroom1_sec = bsync.Sections.Section(ID="Classroom-Pod1")
classroom1_sec += bsync.SectionType('Space function')
classroom1_sec += bsync.OccupancyClassification("Classroom")
classroom1_sec += bsync.OriginalOccupancyClassification("Education-Primary")
sections += classroom1_sec

# add the floor area

# 5.3.4.a Gross floor area for the section (we also add conditioned for 5.3.4.f)
classroom1_fas = bsync.FloorAreas(
    bsync.FloorArea(
        bsync.FloorAreaType("Gross"),
        bsync.FloorAreaValue(12400.02)
    ),
    bsync.FloorArea(
        bsync.FloorAreaType("Conditioned"),
        bsync.FloorAreaValue(12400.02)
    )
)

# 5.3.4.b and 5.3.4.c are asking for typical operating hours for the section
classroom1_tous = bsync.TypicalOccupantUsages()
classroom1_tou_hpw = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Hours per week"),
    bsync.TypicalOccupantUsageValue(45.0)
)
classroom1_tou_wpy = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Weeks per year"),
    bsync.TypicalOccupantUsageValue(50.)
)
classroom1_tous += classroom1_tou_hpw
classroom1_tous += classroom1_tou_wpy

# 5.3.4.d looks for normal / design or peak occupancy.

> # Here, we take values from the IDF as the building has been modelled: with no occupants in the mech room, bathroom or corridors.

Reviewer: ```suggestion
# Here, we take values from the IDF as the building has been modeled: with no occupants in the mech room, lobby, bathroom or corridors.
```

classroom2_sec = bsync.Sections.Section(ID="Classroom-Pod2")
classroom2_sec += bsync.SectionType('Space function')
classroom2_sec += bsync.OccupancyClassification("Classroom")
classroom2_sec += bsync.OriginalOccupancyClassification("Education-Primary")
sections += classroom2_sec

# add the floor area

# 5.3.4.a Gross floor area for the section (we also add conditioned for 5.3.4.f)
classroom2_fas = bsync.FloorAreas(
    bsync.FloorArea(
        bsync.FloorAreaType("Gross"),
        bsync.FloorAreaValue(12400.02)
    ),
    bsync.FloorArea(
        bsync.FloorAreaType("Conditioned"),
        bsync.FloorAreaValue(12400.02)
    )
)

# 5.3.4.b and 5.3.4.c are asking for typical operating hours for the section
classroom2_tous = bsync.TypicalOccupantUsages()
classroom2_tou_hpw = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Hours per week"),
    bsync.TypicalOccupantUsageValue(45.0)
)
classroom2_tou_wpy = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Weeks per year"),
    bsync.TypicalOccupantUsageValue(50.)
)
classroom2_tous += classroom2_tou_hpw
classroom2_tous += classroom2_tou_wpy

# 5.3.4.d looks for normal / design or peak occupancy.

> # Here, we take values from the IDF as the building has been modelled: with no occupants in the mech room, bathroom or corridors.

Reviewer: ```suggestion
# Here, we take values from the IDF as the building has been modeled: with no occupants in the mech room, lobby, bathroom or corridors.
```

classroom3_sec = bsync.Sections.Section(ID="Classroom-Pod3")
classroom3_sec += bsync.SectionType('Space function')
classroom3_sec += bsync.OccupancyClassification("Classroom")
classroom3_sec += bsync.OriginalOccupancyClassification("Education-Primary")
sections += classroom3_sec

# add the floor area

# 5.3.4.a Gross floor area for the section (we also add conditioned for 5.3.4.f)
classroom3_fas = bsync.FloorAreas(
    bsync.FloorArea(
        bsync.FloorAreaType("Gross"),
        bsync.FloorAreaValue(10656.27)
    ),
    bsync.FloorArea(
        bsync.FloorAreaType("Conditioned"),
        bsync.FloorAreaValue(10656.27)
    )
)

# 5.3.4.b and 5.3.4.c are asking for typical operating hours for the section
classroom3_tous = bsync.TypicalOccupantUsages()
classroom3_tou_hpw = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Hours per week"),
    bsync.TypicalOccupantUsageValue(45.0)
)
classroom3_tou_wpy = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Weeks per year"),
    bsync.TypicalOccupantUsageValue(50.)
)
classroom3_tous += classroom3_tou_hpw
classroom3_tous += classroom3_tou_wpy

# 5.3.4.d looks for normal / design or peak occupancy.

> # Here, we take values from the IDF as the building has been modelled: with no occupants in the mech room, bathroom or corridors.

Reviewer: ```suggestion
# Here, we take values from the IDF as the building has been modeled: with no occupants in the mech room, lobby, bathroom or corridors.
```

# create a new section
corridor1_sec = bsync.Sections.Section(ID="Corridor-Pod1")
corridor1_sec += bsync.SectionType('Space function')
corridor1_sec += bsync.OccupancyClassification("Corridor")
corridor1_sec += bsync.OriginalOccupancyClassification("Education-Primary")
sections += corridor1_sec

# add the floor area

# 5.3.4.a Gross floor area for the section (we also add conditioned for 5.3.4.f)
corridor1_fas = bsync.FloorAreas(
    bsync.FloorArea(
        bsync.FloorAreaType("Gross"),
        bsync.FloorAreaValue(2066.67)
    ),
    bsync.FloorArea(
        bsync.FloorAreaType("Conditioned"),
        bsync.FloorAreaValue(2066.67)
    )
)

# 5.3.4.b and 5.3.4.c are asking for typical operating hours for the section
corridor1_tous = bsync.TypicalOccupantUsages()
corridor1_tou_hpw = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Hours per week"),
    bsync.TypicalOccupantUsageValue(45.0)
)
corridor1_tou_wpy = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Weeks per year"),
    bsync.TypicalOccupantUsageValue(50.)
)
corridor1_tous += corridor1_tou_hpw
corridor1_tous += corridor1_tou_wpy

# 5.3.4.d looks for normal / design or peak occupancy.

> # Here, we take values from the IDF as the building has been modelled: with no occupants in the mech room, bathroom or corridors.

Reviewer: ```suggestion
# Here, we take values from the IDF as the building has been modeled: with no occupants in the mech room, lobby, bathroom or corridors.
```

# create a new section
corridor2_sec = bsync.Sections.Section(ID="Corridor-Pod2")
corridor2_sec += bsync.SectionType('Space function')
corridor2_sec += bsync.OccupancyClassification("Corridor")
corridor2_sec += bsync.OriginalOccupancyClassification("Education-Primary")
sections += corridor2_sec

# add the floor area

# 5.3.4.a Gross floor area for the section (we also add conditioned for 5.3.4.f)
corridor2_fas = bsync.FloorAreas(
    bsync.FloorArea(
        bsync.FloorAreaType("Gross"),
        bsync.FloorAreaValue(2066.67)
    ),
    bsync.FloorArea(
        bsync.FloorAreaType("Conditioned"),
        bsync.FloorAreaValue(2066.67)
    )
)

# 5.3.4.b and 5.3.4.c are asking for typical operating hours for the section
corridor2_tous = bsync.TypicalOccupantUsages()
corridor2_tou_hpw = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Hours per week"),
    bsync.TypicalOccupantUsageValue(45.0)
)
corridor2_tou_wpy = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Weeks per year"),
    bsync.TypicalOccupantUsageValue(50.)
)
corridor2_tous += corridor2_tou_hpw
corridor2_tous += corridor2_tou_wpy

# 5.3.4.d looks for normal / design or peak occupancy.

> # Here, we take values from the IDF as the building has been modelled: with no occupants in the mech room, bathroom or corridors.

Reviewer: ```suggestion
# Here, we take values from the IDF as the building has been modeled: with no occupants in the mech room, lobby, bathroom or corridors.
```

corridormain_sec += bsync.SectionType('Space function')
corridormain_sec += bsync.OccupancyClassification("Corridor")
corridormain_sec += bsync.OriginalOccupancyClassification("Education-Primary")
sections += corridormain_sec

# add the floor area

# 5.3.4.a Gross floor area for the section (we also add conditioned for 5.3.4.f)
corridormain_fas = bsync.FloorAreas(
    bsync.FloorArea(
        bsync.FloorAreaType("Gross"),
        bsync.FloorAreaValue(2066.67)
    ),
    bsync.FloorArea(
        bsync.FloorAreaType("Conditioned"),
        bsync.FloorAreaValue(2066.67)
    )
)

# 5.3.4.b and 5.3.4.c are asking for typical operating hours for the section
corridormain_tous = bsync.TypicalOccupantUsages()
corridormain_tou_hpw = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Hours per week"),
    bsync.TypicalOccupantUsageValue(45.0)
)
corridormain_tou_wpy = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Weeks per year"),
    bsync.TypicalOccupantUsageValue(50.)
)
corridormain_tous += corridormain_tou_hpw
corridormain_tous += corridormain_tou_wpy

# 5.3.4.d looks for normal / design or peak occupancy.

> # Here, we take values from the IDF as the building has been modelled: with no occupants in the mech room, bathroom or corridors.

Reviewer: ```suggestion
# Here, we take values from the IDF as the building has been modeled: with no occupants in the mech room, lobby, bathroom or corridors.
```

```

#### 5.2.11 Lobby

```python
# create a new section
lobby_sec = bsync.Sections.Section(ID="Lobby")
lobby_sec += bsync.SectionType('Space function')
lobby_sec += bsync.OccupancyClassification("Lobby")
lobby_sec += bsync.OriginalOccupancyClassification("Education-Primary")
sections += lobby_sec

# add the floor area

# 5.3.4.a Gross floor area for the section (we also add conditioned for 5.3.4.f)
lobby_fas = bsync.FloorAreas(
    bsync.FloorArea(
        bsync.FloorAreaType("Gross"),
        bsync.FloorAreaValue(1840.62)
    ),
    bsync.FloorArea(
        bsync.FloorAreaType("Conditioned"),
        bsync.FloorAreaValue(1840.62)
    )
)

# 5.3.4.b and 5.3.4.c are asking for typical operating hours for the section
lobby_tous = bsync.TypicalOccupantUsages()
lobby_tou_hpw = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Hours per week"),
    bsync.TypicalOccupantUsageValue(45.0)
)
lobby_tou_wpy = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Weeks per year"),
    bsync.TypicalOccupantUsageValue(50.)
)
lobby_tous += lobby_tou_hpw
lobby_tous += lobby_tou_wpy

# 5.3.4.d looks for normal / design or peak occupancy.

> # Here, we take values from the IDF as the building has been modelled: with no occupants in the mech room, bathroom, lobby or corridors.

Reviewer: ```suggestion
# Here, we take values from the IDF as the building has been modeled: with no occupants in the mech room, lobby, bathroom or corridors.
```

```python
# create a new section
wb_sec = bsync.Sections.Section(ID="Whole-Building")
wb_sec += bsync.SectionType('Whole building')
wb_sec += bsync.OccupancyClassification("Education-Primary")
#wb_sec += bsync.OriginalOccupancyClassification("Education-Primary")
sections += wb_sec

# add the floor area

# 5.3.4.a Gross floor area for the section (we also add conditioned for 5.3.4.f)
wb_fas = bsync.FloorAreas(
    bsync.FloorArea(
        bsync.FloorAreaType("Gross"),
        bsync.FloorAreaValue(73960.)
    ),
    bsync.FloorArea(
        bsync.FloorAreaType("Conditioned"),
        bsync.FloorAreaValue(73960.)
    )
)

# 5.3.4.b and 5.3.4.c are asking for typical operating hours for the section
wb_tous = bsync.TypicalOccupantUsages()
wb_tou_hpw = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Hours per week"),
    bsync.TypicalOccupantUsageValue(45.0)
)
wb_tou_wpy = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Weeks per year"),
    bsync.TypicalOccupantUsageValue(50.)
)
wb_tous += wb_tou_hpw
wb_tous += wb_tou_wpy

# 5.3.4.d looks for normal / design or peak occupancy.

> # Here, we take values from the IDF as the building has been modelled: with no occupants in the mech room, bathroom, lobby or corridors.

Reviewer: ```suggestion
# Here, we take values from the IDF as the building has been modeled: with no occupants in the mech room, lobby, bathroom or corridors.
```

wb_occ_levels = bsync.OccupancyLevels()
wb_occ_design = bsync.OccupancyLevel(
    bsync.OccupantQuantityType("Peak total occupants"),
    bsync.OccupantQuantity(1849.)
)
wb_occ_levels += wb_occ_design

# finally, we add these to the actual section element
wb_sec += wb_fas
wb_sec += wb_tous
wb_sec += wb_occ_levels

> ```

Reviewer: ```suggestion
pretty_print(sections)
```

light_systems = bsync.LightingSystems()
systems += light_systems
for section in plugloads:
    lp_ls = bsync.LinkedSectionID(IDref=section['spaceid'])
    ls = bsync.LightingSystem(
        bsync.LampType(
            bsync.LinearFluorescent(
                bsync.LinearFluorescent.LampLabel("T8")
            )
        ),
        bsync.BallastType("Standard Electronic"),
        bsync.LinkedPremises(
            bsync.LinkedPremises.Section(
                lp_ls
            )
        ),
        ID=f"LightingSystem-{section['spaceid']}"
    )
    globals()[f"{section['spaceid'].lower().replace('-', '_')}_lights_lp"] = lp_ls
    globals()[f"lights_{section['spaceid'].lower().replace('-', '_')}"] = ls
    light_systems += ls
```

```python
# The Primary School prototype model has 3 packaged rooftop units (single-speed dx cooling and fuel fired furnace) and 4 packaged rooftop VAV with hot water reheat.
# We create them and link them back to the sections of interest.
# At this point, we aren't modeling separate HVAC zones.
# This level of detail fulfills requirements for a Level 1 audit

hvac_systems = bsync.HVACSystems()
systems+=hvac_systems

> plugloads = [{'spaceid' : lobby_sec["ID"], 'plugload' : 4.0},

Reviewer: do we need to define plugloads again?

# Now, the packaged single-zone constant air volume rooftop units with multi speed DX coil and fuel-fired furnace

# For the first PSZ-AC:
pszac_gym_ls = bsync.LinkedSectionID(IDref=gym_sec["ID"])
pszac_1 = bsync.HVACSystem(
        bsync.PrincipalHVACSystemType("Packaged Rooftop Air Conditioner"),
        bsync.LinkedPremises(bsync.LinkedPremises.Section(pszac_gym_ls)),
        ID="PSZAC-Gym"
    )

# For the second PSZ-AC:
pszac_kitchen_ls = bsync.LinkedSectionID(IDref=kitchen_sec["ID"])
pszac_2 = bsync.HVACSystem(
        bsync.PrincipalHVACSystemType("Packaged Rooftop Air Conditioner"),
        bsync.LinkedPremises(bsync.LinkedPremises.Section(pszac_kitchen_ls)),
        ID="PSZAC-Kitchen"
    )

# For the last PSZ-AC:
pszac_cafeteria_ls = bsync.LinkedSectionID(IDref=cafeteria_sec["ID"])
pszac_3 = bsync.HVACSystem(
        bsync.PrincipalHVACSystemType("Packaged Rooftop Air Conditioner"),
        bsync.LinkedPremises(bsync.LinkedPremises.Section(pszac_cafeteria_ls)),
        ID="PSZAC-Cafeteria"
    )

hvac_systems += vav_pod1
hvac_systems += vav_pod2
hvac_systems += vav_pod3
hvac_systems += vav_other
hvac_systems += pszac_1
hvac_systems += pszac_2
hvac_systems += pszac_3

>     

Reviewer: ```suggestion
    pretty_print(systems)
```

- 6.1.2

Now that we have a quick sense of the building, let's start off by looking at the requirements for an ASHRAE 211 Level 1 audit. This leads us to the concept of a [Scenario](https://buildingsync.net/schema/v2.3.0/documentation/BuildingSync_xsd.html#ScenarioType). A Scenario in BuildingSync is used to refer to energy and timeseries data associated with a particular, well, scenario. Specifically, there are 5 primary Scenarios used in BuildingSync, all which relate back to Standard 211:

![Scenario Types](./img/b1-sc-docs.png)

We will repeatedly come back to the concept of a Scenario, as they are core to organizing information in a BuildingSync document. 

The current building measured scenario is intended to capture true measured historical data. Typically this refers to utility bill data, but AMI type data can also be captured (later).  We start off by creating a new scenario element and defining its type as follows:
- `Scenario[ScenarioType/CurrentBuilding/CalculationMethod/Measured]`. This is an XPath expression that can be interpreted as "A Scenario that has the child elements ScenarioType/CurrentBuilding/CalculationMethod/Measured".  The XML for this would look like:

> 

Reviewer: ```suggestion
```xml
```

<Scenario>
    <ScenarioType>
        <CurrentBuilding>
            <CalculationMethod>
                <Measured/>
            </CalculationMethod>
        </CurrentBuilding>
    </ScenarioType>
</Scenario>

> 

Reviewer: ```suggestion
```
```

            bsync.LinkedBuildingID(IDref=b1["ID"])
        )
    ),
    ID="Scenario-POM-LEDs"
)
```

```python
scenarios += sc_1
```

## 5.8 Level 1 Audit Validation

So did what we just went through actually work? Do we have the required information for a Level 1 Audit? Head over to the [use case validator](https://buildingsync.net/validator) to find out!

![Selection Tool](./img/UC-Selection.png)

Use the line below to write the file to disk

```python
bsync_dump(root, file='Reference-PrimarySchool-L100-Audit.xml')
```

    True

Now upload the file using the BSync use case validator. You should see a green check mark for the L100 AUDIT use case!

![Valid](./img/L100ValidPrimarySchool.png)

## 6. Level 2 Audit

Now that we have set up the level 1 audit, we can move on to level 2. Great!

Level 2 audits include all elements from level 1 audits, and add quite a few details to them. Let us walk through the report requirements in order as they appear in ASHRAE Standard 211, Section 6.2.

### 6.1 Facility Description

For section 6.2.1.1, Building Information, we can use the information we collected during the level 1 audit and enrich it somewhat. 

> We will add schedules first (occupancy, lighting, process and plug loads, and equipment). 

Reviewer: ```suggestion
We will add schedules first (occupancy, lighting, equipment and HVAC). 
```

    ID="Doors"
)

windows = bsync.FenestrationSystem(
    bsync.FenestrationType(
        bsync.Window(
            bsync.WindowLayout("Continuous"),
            bsync.WindowSillHeight(3.6),
            bsync.ExteriorShadingType("None"),
            bsync.InteriorShadingType("None")
        )
    ),
    bsync.FenestrationFrameMaterial("Clad"),
    bsync.FenestrationOperation(True),
    bsync.GlassType("Tinted plus low e"),
    bsync.FenestrationGasFill("Air"),
    bsync.FenestrationGlassLayers("Double pane"),
    bsync.FenestrationUFactor(0.571),
    ID="Windows"
)

window_sys += doors
window_sys += windows
systems += window_sys
```

###### 6.2.3.3 Foundations

We define the building foundations:

```python
foundation_sys = bsync.FoundationSystems()
foundation = bsync.FoundationSystem(
    bsync.GroundCouplings(
        bsync.GroundCoupling(
            bsync.SlabOnGrade(
                bsync.SlabInsulationOrientation("None"),
                bsync.SlabHeating("Unheated")
            )
        )
    ),
    bsync.FloorCovering("Carpet"),
    bsync.FloorConstructionType("Built up"),
    ID="Foundation"
)
foundation_sys += foundation
systems += foundation_sys
```

> 

Reviewer: Do we need to add in overall enclosure tightness as well?

Author: I have that in the AirInfiltrationSystems near the end. Do you mean that?

Reviewer: Ah ok - I did not realize it was the same thing,  Thanks for clarifying.

> ###### 6.2.3.4 Ceilings

Reviewer: Didn't see this in 211?

```python
vav_pod1_plants = bsync.Plants()
vav_pod1 += vav_pod1_plants

vav_pod1_plants += bsync.HeatingPlants(
    bsync.HeatingPlant(
        bsync.HeatingPlantCondition("Good"),
        bsync.Boiler(
            bsync.BoilerType('Hot water'),
            bsync.BurnerControlType("On Off"),
            bsync.BurnerQuantity(1),
            bsync.InputCapacity(431299.),
            bsync.Boiler.Capacity(323474.),
            bsync.CapacityUnits('W'),
            bsync.AnnualHeatingEfficiencyValue(0.75),
            bsync.AnnualHeatingEfficiencyUnits('COP'),
            bsync.ThermalEfficiency(0.8),
            bsync.BoilerLWT(155.),
            bsync.CondensingOperation("Unknown"),
            bsync.HeatingStaging("Variable"),
            bsync.CombustionEfficiency(1.),
            bsync.Quantity(1)
        ),
        bsync.HeatingPlant.ControlSystemTypes(
            bsync.ControlSystemType(
                bsync.ControlSystemType.Other(
                    bsync.OtherCommunicationProtocolName("Unknown")
                )
            )
        ),
        bsync.YearInstalled(2018),
        bsync.PrimaryFuel('Natural gas'),

>         bsync.BuildingAutomationSystem(Tr

Reviewer: should we see more of the VAV output here?

Reviewer: Got it! thanks!

>         bsync.BuildingAutomationSystem(Tr

Reviewer: Getting error message: "TypeError: integer expected"

Author: On my end everything works fine, can you tell me which particular line is expecting an integer please?

![Screen Shot 2022-06-29 at 2 20 23 PM](https://user-images.githubusercontent.com/56871007/176537388-25452733-7e3d-454e-a968-a6bb425387d8.png)

Reviewer: Line 8067: "pom_led += bsync.AnnualDemandSavingsCost(872.57)" is where I got the error.  If i change it to 872 it worked.

Author: I see it now! Sorry, somehow it was showing this comment for lines 6666-6669.

3) Adding the information required for a Level 2 energy audit

4) Generating a BuildingSync file and verifying it against version 2.4.0 of the BuildingSync schema

But first, let us import the required libraries and set up some useful functions:

```python
from bsyncpy import bsync
from lxml import etree
from datetime import datetime, date
import eeweather

def pretty_print(element):
    """Simple printing of an xml element from the bsync library"""
    print(etree.tostring(element.toxml(), pretty_print=True).decode('utf-8'))
    
def bsync_dump(root_element, file="example1.xml"):
    """Write the element to the specified file"""
    doctype = '<?xml version="1.0" encoding="UTF-8"?>'
    as_etree = root_element.toxml()
    as_etree.set("xmlns", "http://buildingsync.net/schemas/bedes-auc/2019")
 

    # Have to manually set the version right now. Align release of bsyncpy to this version.
    as_etree.set("version", "2.4.0")  
    output = etree.tostring(as_etree, doctype=doctype, pretty_print=True)
    with open(file, 'wb+') as f:
        f.write(output)
        return True
```

## 4. Setting up the root of the BuildingSync file

> We are going to define the basic building blocks (ha!) of our representation of the primary school prototype building. For that, we need to define a buiding, which belongs to a site, which belongs to a facility, which belongs to the highest-level of our BuildingSync file: the BuildingSync root. It will look like this:

Reviewer 2: haha!

<BuildingSync>
  <Facilities>
    <Facility ID="My-Nice-Facility">
      <Sites>
        <Site ID="My-Beautiful-Site">
          <Buildings>
            <Building ID="My-Awesome-Building">
            </Building>
          </Buildings>
        </Site>
      </Sites>
    </Facility>
  </Facilities>
</BuildingSync>
For that, we will:
- define the BuildingSync root
- add a Facilities entity to it - as BuildingSync lets us define several facilities at once, should we need to
- add a Facility to our Facilities
- define a Sites entity for our Facility - because a Facility might have several sites
- define a Site within our Sites
- define a Buildings entity for our site - again, we might have several buildings
- finally, define the Primary School building

```python
# Defining the root is as easy as that
xmlargs = {
    'xmlns:xsi' : "http://www.w3.org/2001/XMLSchema-instance",
    'xsi:SchemaLocation' : "http://buildingsync.net/schemas/bedes-auc/2019 ../../../BuildingSync.xsd"
}

> # xmlns:xsi = "http://www.w3.org/2001/XMLSchema-instance"

Reviewer 2: ```suggestion
```

    # xsi:SchemaLocation = "http://buildingsync.net/schemas/bedes-auc/2019 ../../../BuildingSync.xsd"

> root = bsync.BuildingSync()#**kwargs=xmlargs)

Reviewer 2: ```suggestion
```

        bsync.FloorAreaValue(1808.33)
    )
)

# 5.3.4.b and 5.3.4.c are asking for typical operating hours for the section
kitchen_tous = bsync.TypicalOccupantUsages()
kitchen_tou_hpw = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Hours per week"),
    bsync.TypicalOccupantUsageValue(45.0)
)
kitchen_tou_wpy = bsync.TypicalOccupantUsage(
    bsync.TypicalOccupantUsageUnits("Weeks per year"),
    bsync.TypicalOccupantUsageValue(50.)
)
kitchen_tous += kitchen_tou_hpw
kitchen_tous += kitchen_tou_wpy

# 5.3.4.d looks for normal / design or peak occupancy.
# Here, we take values from the IDF as the building has been modeled: with no occupants in the mech room, lobby, bathroom or corridors.
kitchen_occ_levels = bsync.OccupancyLevels()
kitchen_occ_design = bsync.OccupancyLevel(
    bsync.OccupantQuantityType("Peak total occupants"),
    bsync.OccupantQuantity(27.)
)
kitchen_occ_levels += kitchen_occ_design

# finally, we add these to the actual section element
kitchen_sec += kitchen_fas
kitchen_sec += kitchen_tous
kitchen_sec += kitchen_occ_levels
```

#### 5.2.5 Gym

```python
# create a new section
gym_sec = bsync.Sections.Section(ID="Gym")

> gym_sec += bsync.SectionType('Space function')

Reviewer 2: ```suggestion
gym_sec += bsync.SectionType("Space function")
```
this repeats...

          <FloorArea>
            <FloorAreaType>Gross</FloorAreaType>
            <FloorAreaValue>73960.000000</FloorAreaValue>
          </FloorArea>
          <FloorArea>
            <FloorAreaType>Conditioned</FloorAreaType>
            <FloorAreaValue>73960.000000</FloorAreaValue>
          </FloorArea>
        </FloorAreas>
      </Section>
    </Sections>
    

### 5.3 Plugs, lighting and HVAC

Level 1 and Level 2 energy audits require information about the primary systems serving a specific section, but with very different degrees of specificity. Taking guidance from the 211 Normative spreadsheet:
- A Level 1 audit just requires high level information
- A Level 2 audit requires doing some more detailed modeling of the actual system of interest (mainly for HVAC).

```python
systems = bsync.Systems()
# Systems, measures, reports and contacts belong to our facility (f1)
f1 += systems

# Let us automate the task this time
# We first create a list of dictionaries that contains information on space IDs and the appropriate plug load. The plug load is taken from the Primary School Prototype ASHRAE Standard 90.1-2016 model
# We take the IDs from the sections we created earlier

> plugloads = [{'spaceid' : lobby_sec["ID"], 'plugload' : 4.0},

Reviewer 2: good job on the space functions! I'm wondering if in the future we can create a dict or something that we store all the space function data in first, then just iterate? Similar to your space plugloads here.

Author: Yes, agreed. I'll also do the same for schedules, it's a bit too dense right now IMHO.
